File: ios/GradeThread/Disclosure/DisclosureTypes.swift
import Foundation

// Defect-disclosure wire models. Mirror the web contract in
// `src/hooks/use-disclosure.ts` and the edge service in
// `services/edge-functions/src/routes/flipdesk-disclosure.ts`. Decoded through
// `EdgeAPI`'s shared decoder (snake_case → camelCase), so server columns map to
// the camelCase properties below automatically.

/// `GET /api/flipdesk/disclosure/item/:itemId`
struct DisclosureData: Decodable, Equatable {
    let graded: Bool
    let item: DisclosureItem?
    let grade: DisclosureGrade?
    let disclosure: DisclosureText?
    let photos: [DisclosurePhoto]?
}

struct DisclosureItem: Decodable, Equatable {
    let id: String
    let title: String?
    let brand: String?
}

struct DisclosureGrade: Decodable, Equatable {
    let overallScore: Double?
    let gradeTier: String?
    let certificateId: String?
}

struct DisclosureText: Decodable, Equatable {
    let plain: String?
    let markdown: String?
    let html: String?
    let defectCount: Int?
    let hasDefects: Bool?
    let styleFeatures: [String]?
}

/// One graded photo with its defect callouts.
struct DisclosurePhoto: Decodable, Equatable, Identifiable {
    let imageType: String
    let url: String
    let annotations: [PhotoAnnotation]

    // image_type repeats across a submission's photos only if the same role was
    // shot twice; combine with the URL for a stable identity.
    var id: String { imageType + "|" + url }
}

/// A numbered defect callout. `bbox` is normalized `[x, y, w, h]` in 0...1, or
/// nil when the defect couldn't be localized (legend-only).
struct PhotoAnnotation: Decodable, Equatable, Identifiable {
    let n: Int
    let issue: String
    let severity: String
    let location: String?
    let bbox: [Double]?

    var id: Int { n }
    var isLocalized: Bool { (bbox?.count ?? 0) == 4 }
}

/// `POST /api/flipdesk/disclosure/item/:itemId/annotated-photo`
struct SaveAnnotatedResponse: Decodable, Equatable {
    let ok: Bool
    let photoId: String?
    let photoUrl: String?
}

/// `POST /api/flipdesk/disclosure/item/:itemId/apply-to-listing`
struct ApplyDisclosureResponse: Decodable, Equatable {
    let applied: Bool
    let alreadyPresent: Bool?
}

File: android/app/src/main/java/com/gradethread/app/disclosure/DisclosureTypes.kt
package com.gradethread.app.disclosure

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * US-1360: defect disclosure — the graded flaws, their photos, and the callouts
 * that mark them.
 */
@Serializable
data class DisclosureData(
    val graded: Boolean = false,
    val item: DisclosureItem? = null,
    val grade: DisclosureGrade? = null,
    val disclosure: DisclosureText? = null,
    val photos: List<DisclosurePhoto> = emptyList(),
)

@Serializable
data class DisclosureItem(
    val id: String = "",
    val title: String? = null,
    val brand: String? = null,
)

@Serializable
data class DisclosureGrade(
    @SerialName("overall_score") val overallScore: Double? = null,
    @SerialName("grade_tier") val gradeTier: String? = null,
    @SerialName("certificate_id") val certificateId: String? = null,
)

@Serializable
data class DisclosureText(
    val plain: String? = null,
    val markdown: String? = null,
    val html: String? = null,
    @SerialName("defect_count") val defectCount: Int? = null,
    @SerialName("has_defects") val hasDefects: Boolean? = null,
)

/** One graded photo and the defects marked on it. */
@Serializable
data class DisclosurePhoto(
    @SerialName("image_type") val imageType: String = "",
    val url: String = "",
    val annotations: List<PhotoAnnotation> = emptyList(),
) {
    /**
     * A role can repeat within one submission, so identity is role + URL.
     * Keying on the role alone would collapse two "detail" shots into one.
     */
    val id: String get() = "$imageType|$url"
}

/**
 * A numbered defect callout.
 *
 * [bbox] is NORMALISED `[x, y, w, h]` in 0..1, or null when the grader couldn't
 * place the defect on the photo. Normalised because the same coordinates have
 * to land in the same spot whether they're drawn on a thumbnail, a full-size
 * composite, or the web — see [DisclosureGeometry].
 */
@Serializable
data class PhotoAnnotation(
    val n: Int = 0,
    val issue: String = "",
    val severity: String = "",
    val location: String? = null,
    val bbox: List<Double>? = null,
) {
    /** Legend-only when the defect has no place on the photo. */
    val isLocalized: Boolean get() = bbox?.size == 4
}

@Serializable
data class SaveAnnotatedResponse(
    val ok: Boolean = false,
    @SerialName("photo_id") val photoId: String? = null,
    @SerialName("photo_url") val photoUrl: String? = null,
)

@Serializable
data class ApplyDisclosureResponse(
    val applied: Boolean = false,
    /**
     * The listing already carried this disclosure. A success, not a no-op
     * failure — and worth saying differently so a seller doesn't push again.
     */
    @SerialName("already_present") val alreadyPresent: Boolean? = null,
)

@Serializable
internal data class AnnotatedPhotoRequest(
    @SerialName("image_type") val imageType: String,
    @SerialName("data_url") val dataUrl: String,
)
